src/libsaml.ts

```diff
@@ -232,8 +234,8 @@ const libSaml = () => {
     * @return {string}
     */
     attributeStatementBuilder(attributes: LoginResponseAttribute[]): string {
-      const attr = attributes.map(({ name, nameFormat, valueTag, valueXsiType }) => {
-        return `<saml:Attribute Name="${name}" NameFormat="${nameFormat}"><saml:AttributeValue xsi:type="${valueXsiType}">{${tagging('attr', valueTag)}}</saml:AttributeValue></saml:Attribute>`;
+      const attr = attributes.map(({ name, nameFormat, valueTag, valueXsiType, valueXmlnsXs, valueXmlnsXsi }) => {
+        return `<saml:Attribute Name="${name}" NameFormat="${nameFormat}"><saml:AttributeValue ${valueXmlnsXs ? `xmlns:xs="${valueXmlnsXs}" ` : ``}${valueXmlnsXsi ? `xmlns:xsi="${valueXmlnsXsi}" ` : ``}xsi:type="${valueXsiType}">{${tagging('attr', valueTag)}}</saml:AttributeValue></saml:Attribute>`;
```

Comment:
```
${valueXmlnsXs ? `xmlns:xs="${valueXmlnsXs}" ` : ``}

${valueXmlnsXsi ? `xmlns:xsi="${valueXmlnsXsi}" ` : ``}
```

There are two extra spaces in the returned value if `valueXmlnsXsi` and `valueXmlnsXs` are specified. Is it done in purpose ? Otherwise, please remove it, then we are good to merge.

Reply:
Yes, It's done in purpose. It won't produce extra spaces if `valueXmlnsXsi` and `valueXmlnsXs `are specified( because there is no space between ${}${} ). However, without inserting  these two spaces in ${} but put space between ${} ${}. When `valueXmlnsXsi` and `valueXmlnsXs `are **NOT** specified, there will be extra spaces.

Reply:
And the tests will fail.

Reply:
@tngan

Reply:
Then can we do the separation outside ? 

```js
${valueXmlnsXs ? `xmlns:xs="${valueXmlnsXs}"` : ``} ${valueXmlnsXsi ? `xmlns:xsi="${valueXmlnsXsi}"` : ``} xsi:type="${valueXsiType}"
```

Reply:
I have tried this, but the tests failed. If `valueXmlnsXs` and `valueXmlnsXsi ` are **not** provided, two extra space will take place.

Reply:
Let me think...

Reply:
```
const attr = attributes.map(({ name, nameFormat, valueTag, valueXsiType, valueXmlnsXs, valueXmlnsXsi }) => {
        valueXmlnsXs = valueXmlnsXs ? `xmlns:xs="${valueXmlnsXs}" ` : ``
        valueXmlnsXsi = valueXmlnsXsi ? `xmlns:xsi="${valueXmlnsXsi}" ` : ``
        return `<saml:Attribute Name="${name}" NameFormat="${nameFormat}"><saml:AttributeValue ${valueXmlnsXs}${valueXmlnsXsi}xsi:type="${valueXsiType}">{${tagging('attr', valueTag)}}</saml:AttributeValue></saml:Attribute>`;
      }).join('');
```

What about this.

Reply:
```js
const optionalNSProps = { 'xmlns:xs': valueXmlnsXs, 'xmlns:xsi': valueXmlnsXsi};
const optionalNSPropsStr = Object.entries(optionalNSProps).filter(([k, v]) => !!v).map(([k, v]) => `${k}="${v}"`).join(' ');

return `... ${optionalNSPropsStr} xsi:type="${valueXsiType}`
```

Reply:
Maybe not `Object.entries`, there should be a polyfill, otherwise, it will break tests for old version of nodejs.

Reply:
```
return `... ${optionalNSPropsStr} xsi:type="${valueXsiType}`
```
One extra space⤴️

```
return `... ${optionalNSPropsStr}xsi:type="${valueXsiType}`
```
No extra space⤴️

Reply:
Instead of polyfill

```js
Object.keys(optionalNSProps).filter(ns => !!optionalNSProps[ns]).map(ns => `${ns}="${optionalNSProps[ns]}"`).join(' ')
```

Reply:
Actually we can include `xsi:type` as well.

Reply:
```
const defaultValueXmlnsXs = 'http://www.w3.org/2001/XMLSchema'
        const defaultValueXmlnsXsi = 'http://www.w3.org/2001/XMLSchema-instance'
        return `<saml:Attribute Name="${name}" NameFormat="${nameFormat}"><saml:AttributeValue xmlns:xs="${valueXmlnsXs ? valueXmlnsXs : defaultValueXmlnsXs}" xmlns:xsi="${valueXmlnsXsi ? valueXmlnsXsi : defaultValueXmlnsXsi}" xsi:type="${valueXsiType}">{${tagging('attr', valueTag)}}</saml:AttributeValue></saml:Attribute>`;
```
![image](https://user-images.githubusercontent.com/27125445/58256616-e8c5e200-7da1-11e9-8703-64527f53bd92.png)

Tests should be changed.
Ive got to go. Talk later.

Reply:
Yes, go ahead.